File: src/main/java/io/vproxy/vfx/ui/shapes/MovableRect.java
package io.vproxy.vfx.ui.shapes;

import io.vproxy.vfx.control.drag.DragHandler;
import io.vproxy.vfx.entity.Rect;
import io.vproxy.vfx.manager.font.FontManager;
import io.vproxy.vfx.manager.font.FontUsages;
import javafx.scene.Cursor;
import javafx.scene.Group;
import javafx.scene.control.Label;
import javafx.scene.paint.Color;
import javafx.scene.shape.Rectangle;

public class MovableRect extends Group {
    private final Rectangle rect;

    public MovableRect(String labelText) {
        rect = new Rectangle();
        rect.setFill(Color.TRANSPARENT);
        rect.setStroke(Color.RED);
        rect.setStrokeWidth(5);
        rect.setCursor(Cursor.MOVE);

        var pointRightBottom = new Rectangle();
        pointRightBottom.setWidth(10);
        pointRightBottom.setHeight(10);
        pointRightBottom.setFill(Color.RED);
        pointRightBottom.setStrokeWidth(0);
        pointRightBottom.setStroke(Color.TRANSPARENT);
        pointRightBottom.setCursor(Cursor.SE_RESIZE);

        rect.widthProperty().addListener((ob, old, now) -> {
            if (now == null) return;
            pointRightBottom.setLayoutX(now.doubleValue() - 10);
        });
        rect.heightProperty().addListener((ob, old, now) -> {
            if (now == null) return;
            pointRightBottom.setLayoutY(now.doubleValue() - 10);
        });

        var label = new Label(labelText) {{
            FontManager.get().setFont(FontUsages.movableShapeLabel, this);
        }};
        label.setTextFill(Color.RED);
        label.setLayoutX(0);
        rect.heightProperty().addListener((ob, old, now) -> {
            if (now == null) return;
            label.setLayoutY(now.doubleValue() + 5);
        });

        var dragHandler = new DragHandler() {
            @Override
            protected void set(double x, double y) {
                setLayoutX(x);
                setLayoutY(y);
            }

            @Override
            protected double[] get() {
                return new double[]{getLayoutX(), getLayoutY()};
            }
        };

        rect.setOnMousePressed(dragHandler);
        rect.setOnMouseDragged(dragHandler);

        var resizeHandler = new DragHandler() {
            @Override
            protected void set(double x, double y) {
                if (x > 5) rect.setWidth(x);
                if (y > 5) rect.setHeight(y);
            }

            @Override
            protected double[] get() {
                return new double[]{rect.getWidth(), rect.getHeight()};
            }
        };

        pointRightBottom.setOnMousePressed(resizeHandler);
        pointRightBottom.setOnMouseDragged(resizeHandler);

        getChildren().addAll(label, rect, pointRightBottom);
    }

    public void from(Rect rect) {
        setLayoutX(rect.x);
        setLayoutY(rect.y);
        setWidth(rect.w);
        setHeight(rect.h);
    }

    public Rect makeRect() {
        var rect = new Rect();
        rect.x = getLayoutX();
        rect.y = getLayoutY();
        rect.w = getWidth();
        rect.h = getHeight();
        return rect;
    }

    public double getWidth() {
        return rect.getWidth();
    }

    public double getHeight() {
        return rect.getHeight();
    }

    public void setWidth(double width) {
        rect.setWidth(width);
    }

    public void setHeight(double height) {
        rect.setHeight(height);
    }
}

File: src-ts/src/ui/shapes/MovableRect.ts
import { Group } from '../../javafx/Pane.js';
import { RectangleShape } from '../../javafx/shapes.js';
import { Label } from '../../javafx/Label.js';
import { Color } from '../../javafx/color.js';
import { Rect } from '../../entity/Rect.js';
import { FontManager } from '../../manager/font/FontManager.js';
import { FontUsages } from '../../manager/font/FontUsages.js';
import { DragHandler } from '../../control/drag/DragHandler.js';

export class MovableRect extends Group {
  private readonly rect: RectangleShape;

  constructor(labelText: string) {
    super();
    this.rect = new RectangleShape();
    this.rect.setFill(Color.TRANSPARENT);
    this.rect.setStroke(Color.RED);
    this.rect.setStrokeWidth(5);
    this.rect.setCursor('move');

    const pointRightBottom = new RectangleShape();
    pointRightBottom.setWidth(10);
    pointRightBottom.setHeight(10);
    pointRightBottom.setFill(Color.RED);
    pointRightBottom.setStrokeWidth(0);
    pointRightBottom.setStroke(Color.TRANSPARENT);
    pointRightBottom.setCursor('se-resize');

    this.rect.widthProperty.addListener((_old, now) => {
      pointRightBottom.setLayoutX(now - 10);
    });
    this.rect.heightProperty.addListener((_old, now) => {
      pointRightBottom.setLayoutY(now - 10);
    });

    const label = new Label(labelText);
    FontManager.get().setFont(FontUsages.movableShapeLabel, label);
    label.setTextFill(Color.RED);
    label.setLayoutX(0);
    this.rect.heightProperty.addListener((_old, now) => {
      label.setLayoutY(now + 5);
    });

    const self = this;
    const dragHandler = new (class extends DragHandler {
      protected set(x: number, y: number): void {
        self.setLayoutX(x);
        self.setLayoutY(y);
      }

      protected get(): [number, number] {
        return [self.getLayoutX(), self.getLayoutY()];
      }
    })();

    this.rect.setOnMousePressed(dragHandler.handlePressed);
    this.rect.setOnMouseDragged(dragHandler.handleDragged);

    const resizeHandler = new (class extends DragHandler {
      protected set(x: number, y: number): void {
        if (x > 5) self.rect.setWidth(x);
        if (y > 5) self.rect.setHeight(y);
      }

      protected get(): [number, number] {
        return [self.rect.getWidth(), self.rect.getHeight()];
      }
    })();

    pointRightBottom.setOnMousePressed(resizeHandler.handlePressed);
    pointRightBottom.setOnMouseDragged(resizeHandler.handleDragged);

    this.getChildren().addAll(label, this.rect, pointRightBottom);
  }

  from(rect: Rect): void {
    this.setLayoutX(rect.x);
    this.setLayoutY(rect.y);
    this.setWidth(rect.w);
    this.setHeight(rect.h);
  }

  makeRect(): Rect {
    const rect = new Rect();
    rect.x = this.getLayoutX();
    rect.y = this.getLayoutY();
    rect.w = this.getWidth();
    rect.h = this.getHeight();
    return rect;
  }

  getWidth(): number {
    return this.rect.getWidth();
  }

  getHeight(): number {
    return this.rect.getHeight();
  }

  setWidth(width: number): void {
    this.rect.setWidth(width);
  }

  setHeight(height: number): void {
    this.rect.setHeight(height);
  }
}
